model/task_queue_service.go

```diff
@@ -0,0 +1,197 @@
+		order:    order,
+		units:    units,
+	}
+	return t
+}
+
+// FindNextTask returns the next dispatchable task in the queue.
+func (t *TaskDispatchService) FindNextTask(spec TaskSpec) *TaskQueueItem {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// If units is empty, the queue has been emptied. Reset order as an optimization so that the
+	// service no longer needs to iterate over it and check each item against the map.
+	if len(t.units) == 0 {
+		t.order = []string{}
+		return nil
+	}
+
+	var unit schedulableUnit
+	var ok bool
+	var next *TaskQueueItem
+
+	if spec.Group != "" {
+		unit, ok = t.units[compositeGroupId(spec.Group, spec.BuildVariant, spec.Version)]
+		if ok {
+			if next = t.nextTaskGroupTask(unit); next != nil {
+				return next
+			}
+		}
+		// If the task group is not present in the task group map, it has been dispatched.
+		// Fall through to getting a task not in that group.
+	}
+	var numHosts int
+	var err error
+	for _, schedulableUnitID := range t.order {
+		unit, ok = t.units[schedulableUnitID]
+		if !ok {
+			continue
+		}
+		// If maxHosts is not set, this is not a task group.
+		if unit.maxHosts == 0 {
+			delete(t.units, schedulableUnitID)
+			return &unit.tasks[0]
+		}
+		if unit.runningHosts < unit.maxHosts {
+			// TODO: For a multi-host task group, it's not possible to correctly
+			// dispatch tasks based on number of hosts running tasks in the task group
+			// without a transaction. When we use a driver that supports transactions,
+			// we likely want to rewrite this code to use transactions. Currently it's
+			// possible to dispatch a multi-host task group to more tasks than max
+			// hosts. Before transactions are available, it might be possible to do this
+			// better by performing this query again, just before returning from this
+			// function.
+			numHosts, err = host.NumHostsByTaskSpec(spec.Group, spec.BuildVariant, spec.ProjectID, spec.Version)
+			if err != nil {
+				grip.Error(message.WrapError(err, message.Fields{
+					"message": "problem running NumHostsByTaskSpec query",
+					"group":   spec.Group,
+					"variant": spec.BuildVariant,
+					"project": spec.ProjectID,
+					"version": spec.Version,
+				}))
+				return nil
+			}
+			unit.runningHosts = numHosts
+			t.units[schedulableUnitID] = unit
+			if unit.runningHosts < unit.maxHosts {
+				if next = t.nextTaskGroupTask(unit); next != nil {
+					return next
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func compositeGroupId(group, variant, version string) string {
+	return fmt.Sprintf("%s-%s-%s", group, variant, version)
+}
+
+func (t *TaskDispatchService) nextTaskGroupTask(unit schedulableUnit) *TaskQueueItem {
+	for i, nextTask := range unit.tasks {
+		if nextTask.IsDispatched == true {
+			continue
+		}
+
+		nextTaskFromDB, err := task.FindOneId(nextTask.Id)
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"message": "problem finding task in db",
+				"task":    nextTask.Id,
+			}))
+			return nil
```

Comment:
We should probably continue rather than returning in these 2 error scenarios, otherwise I think a task group can get stuck if there is bad data

Reply:
I would prefer that it gets stuck. I'm concerned about the alternative, that the database is unavailable for a moment, and as a result task group dispatching does not handle dependencies correctly.

Reply:
That makes sense for this error case, but what about the one right after it? Couldn't this block all task dispatching?

Reply:
I don't think so. It's either called from https://github.com/evergreen-ci/evergreen/pull/2071/files/c11c31cc981f05a72dda21333d30b7425647d5fb#diff-0fc4657288116846555f63f234c136b7R99, which will fall through to this loop https://github.com/evergreen-ci/evergreen/pull/2071/files/c11c31cc981f05a72dda21333d30b7425647d5fb#diff-0fc4657288116846555f63f234c136b7R108, or it's called in the loop, which means that returning nil will simply move to the next unit in the order.